### PR TITLE
Change to fetch container running status rather than using the cached…

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

In `upgrade-charm` process, the pod will be re-created if the k8s spec changed.
Because the container running status is refreshed when the operator gets watcher event triggered by the state, this might happen before the container is recreated.
Thus, the remote-init process might be execing to the newly created container using the running status of the previously terminated container.
In this PR, instead of using the cached running status in the uniter watcher's remote state, we fetch the current running status before starting the remote-init process. We also retry the remote-init process if we got a container not running error.

## QA steps

```console
## deploy mariab db;
$ juju deploy /tmp/charm-builds/mariadb-k8s/ --debug  --resource mysql_image=mariadb --storage database=100M

$ juju scale-application mariadb-k8s -m t1 5

## change k8s spec then re-build the charm and run upgrade-charm;
$ juju upgrade-charm mariadb-k8s --path /tmp/charm-builds/mariadb-k8s/ --debug --resource mysql_image=mariadb --storage database=100M

## watch the status of pods to see if they are rolling upgraded peacefully.
$ mkubectl get all -o wide

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1877935

### CODE REVIEW

File: worker/caasoperator/initializer.go [hunk 1]
```diff
@@ -107,29 +107,26 @@ func (p initializeUnitParams) Validate() error {
...
-		MaxDuration: 30 * time.Second,
-		Clock:       clk,
-		Stop:        cancel,
-		Func:        do,
-		IsFatalError: func(err error) bool {
-			return err != nil && !exec.IsExecRetryableError(err)
-		},
+		Attempts:     5,
+		Delay:        1 * time.Second,
+		MaxDuration:  30 * time.Second,
```
Comment (wallyworld) [inline worker/caasoperator/initializer.go:125]: I think we may need to consider using different values depending on what's being run. ie pass in as params
We can do some tests on AKS, GKE, microk8s under load etc to see

File: worker/caasoperator/caasoperator.go [hunk 2]
```diff
@@ -582,7 +582,21 @@ func (op *caasOperator) loop() (err error) {
 						return op.runningStatus(unitTag, providerID)
 					}
 					params.RemoteInitFunc = func(runningStatus uniterremotestate.ContainerRunningStatus, cancel <-chan struct{}) error {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply worker/caasoperator/caasoperator.go:584]: needed for now, can fix later

File: worker/caasoperator/initializer.go (hunk 1 above)
Comment (ycliuhw) [reply worker/caasoperator/initializer.go:125]: the point to have this func is to share the retry config for retries inside remote process.
I don't think we need a param here or we just remove this func and call `retry.Call(retry.CallArgs{})` anywhere directly.

File: worker/caasoperator/caasoperator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ycliuhw) [reply worker/caasoperator/caasoperator.go:584]: we need get the `runningStatus.PodName` from there.

**Comment by ycliuhw** [comment]
Tested on `statelss`, `stateful` and `daemon` application.
$$merge$$

**Comment by ycliuhw** [comment]
14:49:28 fatal: unable to access 'https://github.com/tebeka/go2xunit/': Could not resolve host: github.com
$$merge$$

**Comment by ycliuhw** [comment]
https://jenkins.juju.canonical.com/job/github-make-check-juju/5663/testReport/junit/github/com_juju_juju_worker_certupdater/TestPackage/
https://jenkins.juju.canonical.com/job/github-make-check-juju/5663/testReport/junit/github/com_juju_juju_worker_raft_raftclusterer/TestPackage/

**Pull request merged**